### PR TITLE
Fix i64-using function pointers in upstream dynamic linking

### LINKED ISSUE #9562: -s MAIN_MODULE=1 + upstream + function pointer calls using i64  =>  TypeError: cannot pass i64 to or from JS
Hi everybody! I'm starting to learn emscripten and ran in the following problem when reading from a preloaded file (input.txt)

```C++
#include <cstdio>

int main() {
    FILE* file_ = std::fopen("input.txt","rb");
    printf("file pointer %p\n",(void*)file_);
    if (file_) {
        std::fseek(file_, 0l, SEEK_END);
        std::fclose(file_);
    }
    return 0;
}
```

when using
`emcc main.cpp -g -s MAIN_MODULE=1 --preload-file input.txt -o main.html -s EXPORTED_FUNCTIONS=['_main']`

on the latest upstream version (atm 1.38.46 (commit 41adfea99dd4af48c9d068f572d42bb6606384f1)).

The runtime error on fseek in firefox is
```
exception thrown: TypeError: cannot pass i64 to or from JS,@http://0.0.0.0:8002/main.js line 1015 > WebAssembly.Module:wasm-function[0]:0x0
__fseeko_unlocked@http://0.0.0.0:8002/main.js line 2316 > WebAssembly.instantiate:wasm-function[1072]:0x33216
__fseeko@http://0.0.0.0:8002/main.js line 2316 > WebAssembly.instantiate:wasm-function[1073]:0x3324f
fseek@http://0.0.0.0:8002/main.js line 2316 > WebAssembly.instantiate:wasm-function[1074]:0x3327f
__original_main@http://0.0.0.0:8002/main.js line 2316 > WebAssembly.instantiate:wasm-function[163]:0x6dec
main@http://0.0.0.0:8002/main.js line 2316 > WebAssembly.instantiate:wasm-function[164]:0x6e19
Module._main@http://0.0.0.0:8002/main.js:40295:32
callMain@http://0.0.0.0:8002/main.js:49234:22
doRun@http://0.0.0.0:8002/main.js:49294:23
run/<@http://0.0.0.0:8002/main.js:49305:7
```
In Chromium I get a similar TypeError...

Compiling with `-s MAIN_MODULE=0`
`emcc main.cpp -g -s MAIN_MODULE=0 --preload-file input.txt -o main.html -s EXPORTED_FUNCTIONS=['_main']`
works as expected.

Anybody any ideas?

Thanks!

### PR DESCRIPTION
**Pull request by kripken**
The dynamic loader needs to place the original wasm functions in the table, so that other modules can call them. The JS-legalized versions will have the wrong types for indirect calls.

To implement this, binaryen's legalization pass exports illegal functions twice, once legalized for JS, and once as `orig$` plus the original name. This PR makes the loader use those.

Fixes #9562 

Depends on https://github.com/WebAssembly/binaryen/pull/2427 to roll first.

### CODE REVIEW

**Comment by kripken** [comment]
The dependencies have all rolled, and looks like tests all pass.

File: emscripten.py [hunk 1]
```diff
@@ -1798,18 +1798,27 @@ def create_fp_accessors(metadata):
                  'perhaps a side module was not linked in? if this symbol was expected to arrive '
                  'from a system library, try to build the MAIN_MODULE with '
                  'EMCC_FORCE_STDLIBS=XX in the environment");')
+    # the name of the original function. this is generally the normal function
+    # name, unless it is legalized, in which case the export is the legalized
+    # version, and the original provided by orig$X
+    original = mangled[1:]
```
Comment (sbc100) [inline emscripten.py]: The original is already called `name` here I think.. it was just mangled to `mangled` above, no?

File: emscripten.py [hunk 2]
```diff
@@ -1798,18 +1798,27 @@ def create_fp_accessors(metadata):
...
 Module['%(full)s'] = function() {
   %(assert)s
-  var func = Module['%(mangled)s'];
-  if (!func)
-    func = %(mangled)s;
+  // Use the wasm function itself, for the table.
+  var func = Module['asm']['%(original)s'];
+  // If there is no wasm function, this may be a JS library function or
+  // something from another module.
+  if (!func) Module['%(mangled)s'];
```
Comment (sbc100) [inline emscripten.py]: This line doesn't seem to do anything.

File: src/support.js [hunk 3]
```diff
@@ -496,10 +499,11 @@ function loadWebAssemblyModule(binary, flags) {
           assert(parts.length == 3)
           var name = parts[1];
           var sig = parts[2];
+          var legalized = sig.indexOf('j') >= 0;
```
Comment (sbc100) [inline src/support.js]: I assume 'j' means i64?

File: src/support.js [hunk 4]
```diff
@@ -436,7 +436,10 @@ function loadWebAssemblyModule(binary, flags) {
     // b) Symbols from loaded modules are not always added to the global Module.
     var moduleLocal = {};
 
-    var resolveSymbol = function(sym, type) {
+    var resolveSymbol = function(sym, type, legalized) {
+      if (legalized) {
+        sym = 'orig$' + sym;
+      }
 #if WASM_BACKEND
       sym = '_' + sym;
```
Comment (sbc100) [inline src/support.js]: Should we use asmjs_mangle here?  Also does the mangling happen before the `orig$` .. i.e. is it `_orig$foo` or `orig$foo`.   Above it looks like your remove the mangled before pretending `orig$`.

File: src/support.js (hunk 3 above)
Comment (kripken) [reply src/support.js]: Yes, `j` is `i64`. I'll add a comment.

File: src/support.js (hunk 4 above)
Comment (kripken) [reply src/support.js]: This is pre-existing code, but I think `asm_mangle` in python makes sense (where it checks if the `_` is needed or not, which it isn't for `dynCall` etc.) while here, all we have are exported symbols at the C level, which does not include things like `dynCall` etc. that are only for JS.

**Comment by kripken** [comment]
Thanks, I pushed updates for the feedback.

**Pull request merged**